src/net/ssl-context.cpp

```diff
@@ -7,43 +7,65 @@
 #include <openssl/ossl_typ.h>
 #include <openssl/ssl.h>
 #include <exception>
+#include <iostream>
+#include <measurement_kit/common.hpp>
 #include "src/net/ssl-context.hpp"
 
 namespace mk {
 namespace net {
 
-SSL_CTX *SslContext::get_client_context() {
-    static SslContext singleton;
-    return singleton.ctx;
-}
-
-SSL *SslContext::get_client_ssl() {
-    SSL *ssl = SSL_new(get_client_context());
+SSL *SslContext::get_client_ssl(std::string hostname) {
+    SSL *ssl = SSL_new(ctx);
     if (ssl == nullptr) {
+        debug("ssl: failed to call SSL_new");
         throw std::exception();
     }
+
+    SSL_set_tlsext_host_name(ssl, hostname.c_str());
     return ssl;
 }
 
-SslContext::SslContext() {
-    // TODO: check return values where applicable
+void SslContext::init(std::string ca_bundle_path) {
     SSL_library_init();
     ERR_load_crypto_strings();
     SSL_load_error_strings();
     OpenSSL_add_all_algorithms();
 
-    // FIXME: the code in here is really simple because we only aim to
-    // smoke test the OpenSSL functionality. What is missing here is at
-    // the minimum code to verify the peer certificate. Also we should
-    // not allow SSLv2 (and probably v3) connections, and we should also
-    // probably set SNI for the other end to respond correctly.
-    ctx = SSL_CTX_new(SSLv23_client_method());
+    debug("ssl: creating ssl context with bundle %s", ca_bundle_path.c_str());
+
+    ctx = SSL_CTX_new(TLSv1_client_method());
     if (ctx == nullptr) {
+        debug("ssl: failed to create SSL_CTX");
         throw std::exception();
     }
+
+    if (SSL_CTX_load_verify_locations(ctx, ca_bundle_path.c_str(), NULL) == 0) {
+        debug("ssl: failed to load verify location");
+        throw std::exception();
+    };
+
+    SSL_CTX_set_verify(ctx, SSL_VERIFY_PEER, NULL);
+}
+
+SslContext::SslContext() {
+    std::string ca_bundle_path;
+    Settings *global_settings = Settings::global();
+    if (global_settings->find("ca_bundle_path") != global_settings->end()) {
+        ca_bundle_path = (*global_settings)["ca_bundle_path"];
+    } else {
+        // XXX should we add other system default locations in here?
```

Comment:
Perhaps. Since for now our focus is on mobile apps, I'd say we can leave the comment in place and we would deal with this problem later in the future.